liquidity/liquidity.go

```diff
@@ -485,6 +571,92 @@ func (m *Manager) makeLoopOutRequest(suggestion *LoopOutRecommendation,
 	}
 }
 
+// worstCaseOutFees calculates the largest possible fees for a loop out swap,
+// comparing the fees for a successful swap to the cost when the client pays
+// the prepay because they failed to sweep the on chain htlc. This is unlikely,
+// because we expect clients to be online to sweep, but we want to account for
+// every outcome so we include it.
+func worstCaseOutFees(prepayRouting, swapRouting, swapFee, minerFee,
+	prepayAmount btcutil.Amount) btcutil.Amount {
+
+	var (
+		successFees = prepayRouting + minerFee + swapFee + swapRouting
+		noShowFees  = prepayRouting + prepayAmount
+	)
+
+	if noShowFees > successFees {
+		return noShowFees
+	}
+
+	return successFees
+}
+
+// existingAutoLoopSummary provides a summary of the existing autoloops which
+// were dispatched during our current budget period.
+type existingAutoLoopSummary struct {
+	// spentFees is the amount we have spent on completed swaps.
+	spentFees btcutil.Amount
+
+	// pendingFees is the worst-case amount of fees we could spend on in
+	// flight autoloops.
+	pendingFees btcutil.Amount
+}
+
+// totalFees returns the total amount of fees that automatically dispatched
+// swaps may consume.
+func (e *existingAutoLoopSummary) totalFees() btcutil.Amount {
+	return e.spentFees + e.pendingFees
+}
+
+// checkExistingAutoLoops calculates the total amount that has been spent by
+// automatically dispatched swaps that have completed, and the worst-case fee
+// total for our set of ongoing, automatically dispatched swaps.
+func (m *Manager) checkExistingAutoLoops(ctx context.Context,
+	loopOuts []*loopdb.LoopOut) (*existingAutoLoopSummary, error) {
+
+	var summary existingAutoLoopSummary
+
+	for _, out := range loopOuts {
+		if out.Contract.Label != labels.AutoOutLabel() {
+			continue
+		}
+
+		// Skip over any swaps that completed before our start time.
+		if out.LastUpdateTime().Before(m.params.AutoFeeStartDate) {
+			continue
+		}
+
+		// If we have a pending swap, we are uncertain of the fees that
+		// it will end up paying. We use the worst-case estimate based
+		// on the maximum values we set for each fee category. This will
+		// likely over-estimate our fees (because we probably won't
+		// spend our maximum miner amount). If a swap is not pending,
+		// it has succeeded or failed so we just record our actual fees
+		// for the swap.
+		if out.State().State.Type() == loopdb.StateTypePending {
+			prepay, err := m.cfg.Lnd.Client.DecodePaymentRequest(
+				ctx, out.Contract.PrepayInvoice,
+			)
+			if err != nil {
+				return nil, err
+			}
+
+			summary.pendingFees += worstCaseOutFees(
+				out.Contract.MaxPrepayRoutingFee,
+				out.Contract.MaxSwapRoutingFee,
+				out.Contract.MaxSwapFee,
+				out.Contract.MaxMinerFee,
+				mSatToSatoshis(prepay.Value),
+			)
+
+		} else {
+			summary.spentFees += out.State().Cost.Total()
```

Comment:
What if the state is `StateTypeFail`?

Reply:
The else will cover `StateTypeFail` and `StateTypeSuccess`, and our costs are reflected in state updates for both?

Reply:
Discussed offline. Please leave a comment as it's not trivial.

Reply:
done, comment at top of block.